lahja/base.py

```diff
@@ -0,0 +1,169 @@
+from abc import (
+    ABC,
+    abstractmethod,
+)
+import logging
+from typing import (  # noqa: F401
+    Any,
+    AsyncGenerator,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    NamedTuple,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from .misc import (
+    BaseEvent,
+    BaseRequestResponseEvent,
+    BroadcastConfig,
+    ConnectionConfig,
+    Subscription,
+)
+from .typing import (
+    AsyncEndpointContextManager,
+)
+
+TResponse = TypeVar('TResponse', bound=BaseEvent)
+TWaitForEvent = TypeVar('TWaitForEvent', bound=BaseEvent)
+TSubscribeEvent = TypeVar('TSubscribeEvent', bound=BaseEvent)
+TStreamEvent = TypeVar('TStreamEvent', bound=BaseEvent)
+
+
+class EndpointAPI(ABC):
+    """
+    The :class:`~lahja.endpoint.Endpoint` enables communication between different processes
+    as well as within a single process via various event-driven APIs.
+    """
+
+    #
+    # Running and Server API
+    #
+    @abstractmethod
+    async def run(self) -> AsyncEndpointContextManager:
+        """
+        Context manager API for running endpoints.
+
+        .. code-block::
+
+            async with endpoint.run() as endpoint:
+                # endpoint running within context
+            # endpoint stopped after
+        """
+        pass
+
+    @classmethod
+    @abstractmethod
+    async def serve(cls, config: ConnectionConfig) -> AsyncEndpointContextManager:
+        """
+        Context manager API for running and endpoint server.
+
+        .. code-block::
+
+            async with endpoint.serve():
+                # server running within context
+            # server stopped
+        """
+        pass
+
+    #
+    # Connection API
+    #
+    @abstractmethod
+    async def connect_to_endpoint(self, config: ConnectionConfig) -> None:
+        """
+        Establish a new connection to an endpoint.
+        """
+        pass
+
+    @abstractmethod
+    def is_connected_to(self, endpoint_name: str) -> bool:
+        """
+        Return whether this endpoint is connected to another endpoint with the given name.
+        """
+        pass
+
+    #
+    # Event API
+    #
+    @abstractmethod
+    async def broadcast(self, item: BaseEvent, config: Optional[BroadcastConfig] = None) -> None:
```

Comment:
This should probably also add an abstract method for `def broadcast_nowait(self, ...)` no?

Reply:
I'm currenty attempting to approach this as `_nowait` APIs being allowed in individual implementations but not part of the core required API.  While I know how to do it in `trio`, it is intentionally not easy and thus I'm trying to avoid those kind of side-effect APIs that toss something into the background and forget about it.

Reply:
Oh you are absolutely right. I forgot for a moment that `trio` is all about getting rid of these "wild" background taskes :) :+1: